### ISSUE TITLE
Index parameter fails on nested collections

### ISSUE THREAD

**Issue by cirrusone** [alirezanet/Gridify#36]
**Describe the bug**
Based on the recent added functionality [GridifyMapper with additional index parameter](https://github.com/alirezanet/Gridify/issues/29), I am getting error `System.NullReferenceException: 'Object reference not set to an instance of an object.'` When attempting this with nested collections.

**To Reproduce**
This example raises the error on line `var expression = gridifyQuery.GetFilteringExpression(gridifyMapper).Compile();`

Code:
```
using Gridify;
using System.Collections.Generic;
using System.Linq;

namespace ConsoleApp1
{
    public static class Program
    {
        public static void Main(string[] args)
        {
            var test = new Test();
            test.Run();
        }
    }

    public class Test
    {
        public void Run()
        {
            List<Level1> level1List = new List<Level1>();
            level1List.Add(new Level1());

            var gridifyMapper = new GridifyMapper<Level1>().GenerateMappings()
                .AddMap("level4_property1", (q, index) => q.Level2List.Select(x => x.Level3.Level4List[index].Property1));

            var gridifyQuery = new GridifyQuery() { Filter = "level4_property1[0] > 5" };
            var expression = gridifyQuery.GetFilteringExpression(gridifyMapper).Compile();

            var query = level1List.Where(expression);
        }
    }

    public class Level1
    {
        public string Name { get; set; }
        public List<Level2> Level2List = new List<Level2>()
        {
            new Level2()
            {
                Name = "1",
                Level3 = new Level3()
                {
                    Name = "2",
                    Level4List = new List<Level4>()
                    {
                        new Level4() {Name = "3", Property1 = 3, Property2 = 4},
                        new Level4() {Name = "4", Property1 = 4, Property2 = 5},
                        new Level4() {Name = "5", Property1 = 5, Property2 = 6}
                    }
                }
            },

            new Level2()
            {
                Name = "6",
                Level3 = new Level3()
                {
                    Name = "7",
                    Level4List = new List<Level4>()
                    {
                        new Level4() {Name = "8", Property1 = 8, Property2 = 9},
                        new Level4() {Name = "9", Property1 = 9, Property2 = 10},
                        new Level4() {Name = "10", Property1 = 10, Property2 = 11}
                    }
                }
            },

        };
    }

    public class Level2
    {
        public string Name { get; set; }
        public Level3 Level3 = new Level3();
    }

    public class Level3
    {
        public string Name { get; set; }
        public List<Level4> Level4List = new List<Level4>();
    }

    public class Level4
    {
        public string Name { get; set; }
        public double Property1 { get; set; }
        public double Property2 { get; set; }
    }
}
```

Is it possible to do this with nested collections?

**Comment by alirezanet** [comment]
Hi, thank you for the feedback.
Testing these edge cases is insanely hard, also I'm not sure yet whether I fixed this issue or not yet. 
can you please confirm that you expected this query to be generated or not? 
```
l1 =>  l1.Level2List.Any(x => (x.Level3.Level4List[0].Property1 > 5))
```

**Comment by alirezanet** [comment]
It should be fixed in `v2.3.3`. 
thank you

**Comment by cirrusone** [comment]
Thank you, it's working well now.

**Comment by cirrusone** [comment]
This bug seems to fixed for List<T> but I get errors with arrays, `System.IndexOutOfRangeException: 'Index was outside the bounds of the array.'` on line ` var expression = gridifyQuery.GetFilteringExpression(gridifyMapper).Compile();`

This isn't just with index or nested parameters.

Example:

```
using Gridify;
using System.Collections.Generic;
using System.Linq;

namespace ConsoleApp1
{
    public static class Program
    {
        public static void Main(string[] args)
        {
            var test = new Test();
            test.Run();
        }
    }

    public class Test
    {
        public void Run()
        {
            List<Level1> level1List = new List<Level1>();
            level1List.Add(new Level1());

            var gridifyMapper = new GridifyMapper<Level1>().GenerateMappings()
                .AddMap("level4_property1", (q, index) => q.Level2List.Select(x => x.Level3.Level4List[index].Property1));

            var gridifyQuery = new GridifyQuery() { Filter = "level4_property1[0] > 5" };
            var expression = gridifyQuery.GetFilteringExpression(gridifyMapper).Compile();

            var query = level1List.Where(expression);
        }
    }

    public class Level1
    {
        public string Name { get; set; }
        public Level2[] Level2List = new Level2[]
        {
            new Level2()
            {
                Name = "1",
                Level3 = new Level3()
                {
                    Name = "2",
                    Level4List = new List<Level4>()
                    {
                        new Level4() {Name = "3", Property1 = 3, Property2 = 4},
                        new Level4() {Name = "4", Property1 = 4, Property2 = 5},
                        new Level4() {Name = "5", Property1 = 5, Property2 = 6}
                    }
                }
            },

            new Level2()
            {
                Name = "6",
                Level3 = new Level3()
                {
                    Name = "7",
                    Level4List = new List<Level4>()
                    {
                        new Level4() {Name = "8", Property1 = 8, Property2 = 9},
                        new Level4() {Name = "9", Property1 = 9, Property2 = 10},
                        new Level4() {Name = "10", Property1 = 10, Property2 = 11}
                    }
                }
            },

        };
    }

    public class Level2
    {
        public string Name { get; set; }
        public Level3 Level3 = new Level3();
    }

    public class Level3
    {
        public string Name { get; set; }
        public List<Level4> Level4List = new List<Level4>();
    }

    public class Level4
    {
        public string Name { get; set; }
        public double Property1 { get; set; }
        public double Property2 { get; set; }
    }
}
```

**Comment by alirezanet** [comment]
This is a different problem that was fixed in `v2.4.4`.
